File: platform/fabric/src/main/java/io/github/bindglam/neko/manager/CommandManager.java
package io.github.bindglam.neko.manager;

import io.github.bindglam.neko.Neko;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.incendo.cloud.SenderMapper;
import org.incendo.cloud.execution.ExecutionCoordinator;
import org.incendo.cloud.fabric.FabricServerCommandManager;
import org.incendo.cloud.permission.Permission;
import org.jetbrains.annotations.NotNull;

import java.util.logging.Logger;

public final class CommandManager implements Managerial {
    @Override
    public void preload(@NotNull Context context) {
        var manager = new FabricServerCommandManager<>(
                ExecutionCoordinator.simpleCoordinator(),
                SenderMapper.identity()
        );

        manager.command(manager.commandBuilder("neko")
                .literal("reload")
                .permission(Permission.of("mint.command.reload"))
                .handler(ctx -> {
                    ctx.sender().sendMessage(Component.text("Reloading...").color(NamedTextColor.YELLOW));
                    Neko.platform().reload();
                    ctx.sender().sendMessage(Component.text("Successfully reloaded!").color(NamedTextColor.GREEN));
                }));

        /*manager.command(manager.commandBuilder("neko")
                .literal("give")
                .permission(Permission.of("mint.command.give"))
                .required("player", PlayerParser.playerParser())
                .required("item", NamespacedKeyParser.namespacedKeyParser())
                .handler(ctx -> {
                    Player player = ctx.get("player");
                    NamespacedKey itemKey = ctx.get("item");

                    var item = RegistryManager.GlobalRegistries.registries().item()
                            .get(itemKey)
                            .orElse(null);
                    if (item == null) {
                        ctx.sender().sendMessage(Component.text("Item not found").color(NamedTextColor.RED));
                        return;
                    }

                    player.getInventory().addItem((ItemStack) item.itemStack().get().unwrap());
                }));*/
    }

    @Override
    public void start(@NotNull Context context) {
    }

    @Override
    public void end(@NotNull Context context) {
    }
}

File: core/src/main/kotlin/com/bindglam/neko/manager/CommandManager.kt
package com.bindglam.neko.manager

import com.bindglam.neko.api.NekoProvider
import com.bindglam.neko.api.content.glyph.GlyphBuilder
import com.bindglam.neko.api.registry.BuiltInRegistries
import com.mojang.brigadier.Command
import com.mojang.brigadier.arguments.IntegerArgumentType
import io.papermc.paper.command.brigadier.Commands
import io.papermc.paper.command.brigadier.argument.ArgumentTypes
import io.papermc.paper.command.brigadier.argument.resolvers.selector.PlayerSelectorArgumentResolver
import io.papermc.paper.plugin.bootstrap.BootstrapContext
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.NamespacedKey

@Suppress("unstableApiUsage")
object CommandManager {
    fun init(context: BootstrapContext) {
        context.lifecycleManager.registerEventHandler(LifecycleEvents.COMMANDS) { commands ->
            Commands.literal("neko")
                .requires { it.sender.isOp }
                .then(Commands.literal("reload")
                    .executes { ctx ->
                        NekoProvider.neko().reload(ctx.source.sender)

                        return@executes Command.SINGLE_SUCCESS
                    }
                )
                .then(Commands.literal("give")
                    .then(Commands.argument("player", ArgumentTypes.player())
                        .then(Commands.argument("key", ArgumentTypes.namespacedKey())
                            .then(Commands.argument("amount", IntegerArgumentType.integer())
                                .executes { ctx ->
                                    val player = ctx.getArgument("player", PlayerSelectorArgumentResolver::class.java).resolve(ctx.source).first()
                                    val key = ctx.getArgument("key", NamespacedKey::class.java)
                                    val amount = ctx.getArgument("amount", Int::class.java)

                                    BuiltInRegistries.ITEMS.get(key).ifPresentOrElse({ customItem ->
                                        player.inventory.addItem(customItem.itemStack().apply { this.amount = amount })
                                        ctx.source.sender.sendMessage(Component.text("Successfully gave an item").color(NamedTextColor.GREEN))
                                    }) {
                                        ctx.source.sender.sendMessage(Component.text("Unknown item key").color(NamedTextColor.RED))
                                    }

                                    return@executes Command.SINGLE_SUCCESS
                                }
                            )
                        )
                    )
                )
                .then(Commands.literal("glyph")
                    .then(Commands.argument("key", ArgumentTypes.namespacedKey())
                        .executes { ctx ->
                            val key = ctx.getArgument("key", NamespacedKey::class.java)

                            BuiltInRegistries.GLYPHS.get(key).ifPresentOrElse({ glyph ->
                                ctx.source.sender.sendMessage(glyph.component(GlyphBuilder.builder().offsetX(0)))
                            }) {
                                ctx.source.sender.sendMessage(Component.text("Unknown glyph key").color(NamedTextColor.RED))
                            }

                            return@executes Command.SINGLE_SUCCESS
                        })
                )
                .also { commands.registrar().register(it.build()) }
        }
    }
}